### ISSUE TITLE
[REVIEW]: Scyland3D: Processing 3D landmarks 

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1262]
**Submitting author:** @ybayle (<a href="http://orcid.org/0000-0002-5143-2335">Yann Bayle</a>)
**Repository:** <a href="https://github.com/ybayle/Scyland3D" target ="_blank">https://github.com/ybayle/Scyland3D</a>
**Version:** v1.1.0
**Editor:** @Kevin-Mattheus-Moerman
**Reviewers:** @r-barnes, @patrikhuber
**Archive:** <a href="https://doi.org/10.5281/zenodo.3637546" target="_blank">10.5281/zenodo.3637546</a>

## Status

[![status](http://joss.theoj.org/papers/4e9dd95a8c973bce0b723a5795a22f58/status.svg)](http://joss.theoj.org/papers/4e9dd95a8c973bce0b723a5795a22f58)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/4e9dd95a8c973bce0b723a5795a22f58"><img src="http://joss.theoj.org/papers/4e9dd95a8c973bce0b723a5795a22f58/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/4e9dd95a8c973bce0b723a5795a22f58/status.svg)](http://joss.theoj.org/papers/4e9dd95a8c973bce0b723a5795a22f58)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/ybayle/Scyland3D" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@r-barnes, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @Kevin-Mattheus-Moerman know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/ybayle/Scyland3D">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v1.1.0
- [x] **Authorship:** Has the submitting author (@ybayle) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/ybayle/Scyland3D">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v1.1.0
- [x] **Authorship:** Has the submitting author (@ybayle) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @r-barnes it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01262/joss.01262/10.21105.joss.01262.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes thanks. This is where the review takes place and you can see the tickboxes at the top here. Note there are also instructions to avoid the spam you mentioned. Let me know if you have questions.

**Comment by Kevin-Mattheus-Moerman** [comment]
@ybayle this is where the review takes place. Please also note the comments by @patrikhuber over at https://github.com/openjournals/joss-reviews/issues/1153

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes are you able to start the review process? Thanks

**Comment by labarba** [comment]
I notice that the `.py` file mentions MIT License, while the `LICENSE` file is GNU. 
https://github.com/ybayle/Scyland3D/blob/1d219d40a8327515d936b94a2d1d4ce7060c0f16/Scyland3D.py#L5

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes can you please confirm you are able to continue with the review process? Thanks

**Comment by ybayle** [comment]
> I notice that the `.py` file mentions MIT License, while the `LICENSE` file is GNU.
> https://github.com/ybayle/Scyland3D/blob/1d219d40a8327515d936b94a2d1d4ce7060c0f16/Scyland3D.py#L5

Thanks for noticing! We have updated our license consistently accordingly.

**Comment by r-barnes** [comment]
ybayle/Scyland3D#1

**Comment by patrikhuber** [comment]
I am moving my comments from the closed thread to here, I think they would be more appropriate here since the other thread is closed.
---
### Overview

It seems to me that in general, this software is for a very, _**very**_ limited use case (which of course isn't necessarily bad!). Basically, the main thing it does:
1) Reads `.pts` landmarks from a software called "Landmark software (v3.6)" and converts them to a `.csv` format
2) It can mirror and re-order those landmarks while converting them. The mirroring potentially seems like it can be a bit tricky (see more below).

### Comments/questions:

* The Readme needs a better introduction in my opinion. It states

> Scyland3D is a Python tool for converting 3D raw landmark and semilandmark coordinates to a usable format for geometric morphometric analyses

but if you're not from _exactly_ that field, you're going to have a hard time discerning if the tool is useful or not, because "3D raw landmarks" is such a large field. I am not from the author's field directly but "geometric morphometric analysis" looks like it is in fact what I know by the term shape analysis, which is a really broad field, with thousands of different landmark formats. And Sciland3D seems only for one specific task in one very particular field (or perhaps even one particular lab or commercial software), without much broader applicability, I think. At least for me as a computer vision researcher, having worked with statistical shape models a lot, I had a hard time figuring out what this exactly is and does.

* The PDF and repo don't explain or introduce what "semilandmarks" are. I've worked with facial landmark points for many years and don't believe I've come across this term. Also no example seems to make use of them, as far as I can tell. Some explanation here would be really great.

* The Readme mentions `.pts` files but not what they are or where they're coming from. We have `.pts` files too in the facial landmarks community and they are something completely different (e.g. see https://github.com/patrikhuber/eos/blob/master/examples/data/image_0010.pts).
The PDF says they're ".pts files, exported by the Landmark software (v3.6)" but it also doesn't help much - please add a link/reference to that "Landmark software (v3.6)" or even better, to the format definition as well.

To sum that part up, I think there's way more context needed, and the name "Scyland3D: Processing 3D landmarks" may also be too general.

### Specific comments regarding the paper (PDF):

* The abstract says "convert them to a usable format". Why would the pts files be a less usable format than csv files? I believe that might be the wrong motivation or the motivation is just misstated. Both formats are probably useful, but data conversion is part of our daily jobs so we just want to to be as easy, painless and error-free as possible.

* I believe "ie" should have periods, i.e. (no pun intended... :-) ) "i.e.". Also I wouldn't italicise it but I'm not sure whether that's perhaps just personal preference. Italicising is for emphasising and what you want to emphasise there is _in situ_ and not the "i.e." :-) (see e.g. https://english.stackexchange.com/questions/3911/correct-spelling-italicization-of-e-g-i-e)

* _"Few free software solutions allow digitizing 3D landmarks and semilandmarks [...]"_: What do you mean with that? Store them to the disk? I think again what you probably mean is that few ready-made scripts exist that convert between all the many formats?

* _"checks and reports [...] Consistency of folder and file names"_ In what way does it check that, what naming scheme does it expect?

* _"checks and reports [...] Precision of landmark and semilandmark coordinates export by the Landmark software."_ - I can't find anything about that in the code?

* _"mirrored objects have reverted bending"_ What do you mean with bending here?

### Specifically with regards to the code:

* How useful is the mirroring? For it to be useful, you usually have to mirror the original data too (in your case e.g. the teeth point-cloud), in the same way. Furthermore it seems to me that e.g. in the example given in Figure 1, this only works if the landmarks are placed in a symmetric way in the first place (i.e. if 1 is the "mirror-point" of 7, etc.).

* The code could use more documentation in general. A tiny bit longer function documentation, but also in-code comments. Code should be either self-explanatory and need no comments (best case) or have appropriate comments, where needed.

* Here it says something about a comparison "up to epsilon": https://github.com/ybayle/Scyland3D/blob/1d219d40a8327515d936b94a2d1d4ce7060c0f16/Scyland3D.py#L80
But in the following code, I can't find any comparison up to an epsilon, only equality checks. It may very well be that I don't read the code correctly and I haven't run it either but it kind of reinforces my point from above about documentation.

* https://github.com/ybayle/Scyland3D/blob/1d219d40a8327515d936b94a2d1d4ce7060c0f16/Scyland3D.py#L106-L107
It would be nice to know here what's "good" about one link and "wrong" about one link. There seem to be some peculiarities at play here, and even a StackOverflow answer with most upvotes and comments like "This one should be marked as the correct answer." that apparently has gotten something important wrong. Let us know about it please, and also again, please put comments in that function and let us know which ones were/are the tricky bits.
From what I can see, that function (`reverse_z`) is actually one of *the* main contributions of the proposed submission.

Some of those comments might be a bit picky but in particular because I don't think this contribution is particularly big, it is even more important that the small thing that *is* contributed is as useful as possible and can be used by people being assured it doesn't have any bugs, or if they do need to check something, the code should be documented clearly enough to make it easy.

Good luck and all the best! :-)

**Comment by r-barnes** [comment]
As far as I can tell there are no tests. Is this a problem?

**Comment by r-barnes** [comment]
I've completed the JOSS rubric and have raised concerns linked above.

**Comment by r-barnes** [comment]
(I want to apologize for the delay in getting to this review, I've been experiencing some issues with RSI which is made it difficult to keep up with computer-based tasks. I'm now using a speech recognition program to do the review, so I apologize for any grammatical  errors or odd choices of words.)

Many of the issues that I've raised are on the target repository.

I share @patrikhuber's concerns that this piece of software is super specific. When I saw the name of the software and description of the softer, I expected to find a large amount of code relating to the manipulation of 3D points, instead the softer it appears to consist predominantly of what's essentially a utility function for a much larger pipeline. Given the limited scope of the software, I would have difficulty recommending it for publication.

Similar to the other reviewer, I don't see any kind of epsilon cutoff to deal with floating-point issues. I also worry about the utility of the mirroring, it seems as though the user has to specify the plane of the mirror. it seems to me that this is a problem that is better suited to a technique such as iterative closest point or optimal transport.

The code is at times verbose:
```
coord = []
    for item in data:
        tmp = []
        for xyz in item.split(","):
            tmp.append(float(xyz))
coord.append(tmp)
```
could be rendered as
```
coord = [float(c) for item in data for c in item.split(",")]
```
Granted, such techniques should not be employed where they impede the readability of code, but I think there are several instances in which the existing code is more complex than is necessary. `export2csv()` might be the best example of this, since Python includes an entire core module for solving exactly this problem.

This contribution is small enough that it should be as perfect and rock-solid as possible. For that, I'm looking for cleaner, better-commented code, test cases, and an exceptionally well-explained use case. Even so, I think this function is just too limited in scope to make a good contribution.

**Comment by r-barnes** [comment]
@Kevin-Mattheus-Moerman

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @r-barnes for your review work and detailed comments!

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @patrikhuber also for your comments here!

**Comment by Kevin-Mattheus-Moerman** [comment]
@ybayle can you formulate a detailed reply to the above comments?

**Comment by ybayle** [comment]
Yes, I have started to fix some issues reported by the reviewers on the main repo! I'll keep you updated.

**Comment by r-barnes** [comment]
Thanks all, do ping me when I'm next needed.

**Comment by r-barnes** [comment]
I notice that I've been assigned. Should I review again?

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes I think we are still waiting for @ybayle to implement changes and work on the issues reported. @ybayle can you provide an update as to where we stand? Thanks!

**Comment by ybayle** [comment]
Sorry for the delay, I am fixing the issues reported and will get back to you ASAP. I just have to finish something else before. I'll ping when everything will be ready!

**Comment by labarba** [comment]
👋  @ybayle — Can you give us an update? If you're not close to done, please let me know of a time period to set a reminder for you.

**Comment by ybayle** [comment]
@labarba: Sorry for the delay, I have taken into account most comments from @r-barnes in the last commits and started processing comments from @patrikhuber locally that will be pushed soon.

**Comment by ybayle** [comment]
@patrikhuber 
Thank you for your valuable comments to improve our submission, we detail below our changes.

> The Readme needs a better introduction in my opinion.
> The PDF and repo don't explain or introduce what "semilandmarks" are.

The ReadMe and the paper have been updated accordingly.

> The Readme mentions .pts files but not what they are or where they're coming from. We have .pts files too in the facial landmarks community and they are something completely different (e.g. see https://github.com/patrikhuber/eos/blob/master/examples/data/image_0010.pts).

As I can see, we provide as examples nearly the same .pts files except, we have an additional column for values for the z-axis. We added a link in the paper and in the ReadMe to the definition of the .pts file format we use.

> "Few free software solutions allow digitizing 3D landmarks and semilandmarks [...]": What do you mean with that? Store them to the disk? I think again what you probably mean is that few ready-made scripts exist that convert between all the many formats?

You raise an interesting question there about the variety of vocabulary used here. We saw in numerous places (e.g. https://www.rdocumentation.org/packages/geomorph/versions/3.1.2/topics/digitize2d or http://people.tamu.edu/~alawing/materials/ESSM689/Quick_Guide_to_Geomorph_v2.0.pdf) the word "digitizing" used for "data acquisition". In order to avoid confusion with other readers we used both words in the paper.

> "mirrored objects have reverted bending" What do you mean with bending here?
> How useful is the mirroring? For it to be useful, you usually have to mirror the original data too (in your case e.g. the teeth point-cloud), in the same way. Furthermore it seems to me that e.g. in the example given in Figure 1, this only works if the landmarks are placed in a symmetric way in the first place (i.e. if 1 is the "mirror-point" of 7, etc.).

We are talking here about the main curvature a surface can display in terms of how much they display a concave or convex shape on a given axis. In our example with shark's teeth, they are longer (x) than wide (y) and very thin (z). Teeth are concave on the z-axis and their tip are not centred relatively to their base. So, when you mirror them to match left teeth to right teeth, you have their base and tip that are overlapping but you one is concave while the other is convex as can be seen on their z-axis. That's why "mirrored objects have reverted bending".

> The code could use more documentation in general. A tiny bit longer function documentation, but also in-code comments. Code should be either self-explanatory and need no comments (best case) or have appropriate comments, where needed.

We had the same comment from the second reviewer and updated our code accordingly.

> Here it says something about a comparison "up to epsilon": https://github.com/ybayle/Scyland3D/blob/1d219d40a8327515d936b94a2d1d4ce7060c0f16/Scyland3D.py#L80 But in the following code, I can't find any comparison up to an epsilon, only equality checks. It may very well be that I don't read the code correctly and I haven't run it either but it kind of reinforces my point from above about documentation.

Indeed, the first version of the script was using an epsilon as a workaround to report duplicated landmarks in the exported .pts files. A better solution has been found since. The comment is now updated.

**Comment by ybayle** [comment]
@r-barnes

> I also worry about the utility of the mirroring, it seems as though the user has to specify the plane of the mirror. it seems to me that this is a problem that is better suited to a technique such as iterative closest point or optimal transport.

We detailed further our code to explain why the mirroring is needed. We also added details about the fact that our code finds a plane and that no user specified plane has to be supplied.

**Comment by ybayle** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01262/joss.01262/10.21105.joss.01262.pdf)

**Comment by patrikhuber** [comment]
Hi @ybayle,

Thanks for addressing the comments! I think it looks much better!

A few things that I think still warrant clarification, or that I've noticed while reviewing the changes:

* > Readme: "...exported from landmark acquisition softwares..."

It still looks to me like your script is very specific, more concretely, it seems specific to the "Landmark 3.6", "Landmark Editor" software from IDAV, UC Davis (http://www.idav.ucdavis.edu/research/EvoMorph).
Sure, other programs may export the exact same .pts files (do you know of any such?), but I think it would clarify the context of your script if this was prominently mentioned in the Readme introduction, and it would help people using the "Landmark 3.6" software to find your script and know that it might be useful to them.

* [In this image](https://github.com/ybayle/Scyland3D/raw/master/paper/figure1.png), don't you want to mirror the landmark numbers as well? Otherwise you change the semantic meaning of each landmark? In my field, when mirroring, we want to preserve the semantic meaning of a landmark when mirroring and not reorder them but it might be different in your use field.

* > "Scyland3D is licensed under the MIT License as described in the license file. Scyland3D was mainly created for research purposes and thus can be used freely for research and academic use with the following citation: [...]"

I'm not sure if the wording of the second sentence could be a bit confusing. It may be interpreted as a restriction on the MIT License? IANAL but perhaps rewording might make sense.

* https://github.com/ybayle/Scyland3D/blob/ba701ad3e73e5a1b0dbf69f8dfb1be4b3dcf8710/Scyland3D.py#L95
"reverts": Not sure I understand. Do you perhaps mean "reverses"?

* https://github.com/ybayle/Scyland3D/blob/ba701ad3e73e5a1b0dbf69f8dfb1be4b3dcf8710/Scyland3D.py#L80
Do you check here if the x, y and z coordinates are the same, or only one of them?

* > "Scyland3D first removes duplicated coordinates because landmarks and
semilandmarks might be mixed up in the Landmark software."

"might be mixed up" is not very clear/scientific. Is this a particular problem/bug in the "Landmark 3.6" software?
It seems to me that if it mixes up landmarks and semilandmarks, that just removing duplicates would not really be a solution?

* Throughout the repo/paper: I think "software" is an uncountable noun and you can't (or shouldn't) use "softwares". In most cases you can just use the singular (or "applications", "pieces of software" etc.). Also there were a couple of instances of dodgy grammar/wording - I'd advise proofreading by a native speaker.

* I think you might be mixing BE/AE? E.g. "characterizing" vs "favoured".

* I found a few more typos across the Readme and Paper (didn't write them down sorry - one of them was "biologist"=>"biologists" in the paper).

**Comment by ybayle** [comment]
Hi @patrikhuber,

Thanks for your comments.

> It still looks to me like your script is very specific, more concretely, it seems specific to the "Landmark 3.6", "Landmark Editor" software from IDAV, UC Davis (http://www.idav.ucdavis.edu/research/EvoMorph).

I have updated the ReadMe to prominently mention this in the introduction.

> In this image, don't you want to mirror the landmark numbers as well? Otherwise you change the semantic meaning of each landmark? In my field, when mirroring, we want to preserve the semantic meaning of a landmark when mirroring and not reorder them but it might be different in your use field.

Indeed, our script handle the inversion or not of the numbering of the landmarks. Our use-case is closely related to this kind of processing, thus justifying our current example image.

> It may be interpreted as a restriction on the MIT License?

I have reworded the license.

> "reverts": Not sure I understand. Do you perhaps mean "reverses"?

Thanks you for pointing out this typo.

> Do you check here if the x, y and z coordinates are the same, or only one of them?

This sentence has been reworded.

> "might be mixed up" is not very clear/scientific.

This has been reworded. Indeed, the landmark acquisition software avoid setting landmarks at the same position but landmarks can be set up at the same position of a semilandmark, which is not wanted.

> Also there were a couple of instances of dodgy grammar/wording. I think you might be mixing BE/AE? E.g. "characterizing" vs "favoured".

The paper and ReadMe have been checked for misspelling and now use BE. Please confirm no mistakes or typos are remaining.

**Comment by ybayle** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01262/joss.01262/10.21105.joss.01262.pdf)

**Comment by patrikhuber** [comment]
Hi @ybayle,

Looks good!

> Indeed, our script handle the inversion or not of the numbering of the landmarks.

I see - but I'm not sure how? Could this be the `order_factor` ? The example in the Readme sets this to `'upper'`. But to be honest I really can't figure out what it might be doing - the code doesn't seem to check for any particular value like 'upper'? What are the valid values, and what do they each do? This could be documented better (e.g. also added to the "help" option).

> `# Reorder the landmarks to avoir bias during landmarks set up by humans`

Typo "avoir" :) Also the meaning is not clear to me - "...during landmarks set up by humans"?

Last thing, `feature_names` is listed in the Readme as an argument to `pts2csv` but a CTRL+F of the code yields no results and it's also not part of the function signature?

**Comment by ybayle** [comment]
Hi @patrikhuber,

> Looks good!

Thank you again for your comments and suggestions that help improve a lot our submission and our code. We also learn many python-related things during this process!

> I see - but I'm not sure how? Could this be the order_factor ? The example in the Readme sets this to 'upper'. But to be honest I really can't figure out what it might be doing - the code doesn't seem to check for any particular value like 'upper'? What are the valid values, and what do they each do? This could be documented better (e.g. also added to the "help" option).

If the dataset contains lower and upper teeth where only the upper teeth need to be reordered, one can use: `Scyland3D.pts2csv("example/", order_factor="upper", order=[36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 37])` or from the command line: `python Scyland3D.py -i "example/" -f "upper" -o "36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 37"`. This is only valid if the keyword `upper` is present in the file name. Information regarding this feature have been added to the ReadMe and to the code.

> Last thing, feature_names is listed in the Readme as an argument to pts2csv but a CTRL+F of the code yields no results and it's also not part of the function signature?

This is fixed and we also added better example in the ReadMe to describe how to use our package from another python script or from the command line.

**Comment by patrikhuber** [comment]
You're welcome, glad it is helpful.

I understand, thanks for the additional explanations.

I think you forgot to document the `feature_names` here in the "Args" list:
https://github.com/ybayle/Scyland3D/blob/3626b2ad4c339ee2f2483d5a0c06dd94df96f615/Scyland3D.py#L151-L152

Also you should format the .py file with PEP 8, there are a few lines that are too long (and potentially some other things). Any Python IDE should be able to do it, in PyCharm it's "Code => Reformat Code".

Otherwise it looks all good from my side! 👍

**Comment by ybayle** [comment]
Hi @patrikhuber,
Everything should be fine now and I have formatted the code using https://github.com/python/black.
Thanks again for your comments!

**Comment by ybayle** [comment]
Hi @r-barnes: the new version v1.0.6 with all the fixes is now available on pypi https://pypi.org/project/Scyland3D/

**Comment by patrikhuber** [comment]
@ybayle Just fyi the tooth image doesn't load on the pypi page, at least in my browser. If you have the problem as well, it might be because of the relative URL. I just checked my project and what I did was use the "raw" URL for the image (I probably found that solution in stackoverflow or somewhere): E.g.: `https://raw.githubusercontent.com/patrikhuber/eos/gh-pages/images/sfm_shape_3448_mesh.png`

**Comment by ybayle** [comment]
Hi @patrikhuber,
The pypi repo now displays correctly the image thanks to your suggested url!

**Comment by ybayle** [comment]
@Kevin-Mattheus-Moerman Could you please update the version to https://github.com/ybayle/Scyland3D/releases/tag/v1.0.16 in the description?

**Comment by r-barnes** [comment]
Sorry. Missed the previous message. I'll get to the review shortly.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set v1.0.16 as version

**Comment by whedon** [comment]
OK. v1.0.16 is the version.

**Comment by Kevin-Mattheus-Moerman** [comment]
@patrikhuber thanks for your comments here. Would you like to be be added as an official reviewer so you have a checklist at the top?

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes thanks for your help with this review. Could you give us an indication as to when you think you can complete the review?

**Comment by patrikhuber** [comment]
@Kevin-Mattheus-Moerman I don't mind personally. But if there's anything more that would help you (like doing an official checklist), I'd be happy to do it. Should be trivial given I've already completed a pretty much full review and everything is fine from my end.
I am on vacation though so the earliest I could do anything is the week of 12 August.

**Comment by r-barnes** [comment]
When I try `pip install -U Scyland3D` and run `Scyland3D` from the CLI, I get
```
Traceback (most recent call last):
  File "/home/rick/.local/bin/Scyland3D", line 7, in <module>
    from Scyland3D import main
ImportError: cannot import name 'main'
```
a similar error occurs when using `python3 setup.py install --user`

**Comment by r-barnes** [comment]
I see the following issues:
 * [ ] Both `pip3 install -U Scyland3D` and `python3 setup.py install Scyland3D --user` lead to an unusable state: `ImportError: cannot import name 'main'`
 * [ ] Calling without command line arguments does not provide help. Instead, an example directory is parsed. This violates the principle of least surprise.
 * [ ] Argument parsing seems unnecessarily verbose. My first review suggested the `argparse` package. I'm not sure why `getopt` has been used.
 * [ ] Image in README appears without justification. It should be used to support the initial description of what the software does.
 * [ ] Internal comments are lacking near critical places. What, for instance, does `t = float((D - A * x - B * y - C * z) / (A * A + B * B + C * C))` do? Try to describe mathematical operations and other patches of non-contextualized code to orientate yourself and others. Particularly, [Issue 8](https://github.com/ybayle/Scyland3D/issues/8) seems disappointingly unaddressed.

Other than this, the situation is improved versus the original submission; however, my sense that the software solves too specific a problem in too specific a way remains.

**Comment by r-barnes** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

```

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon add @patrikhuber as reviewer

**Comment by whedon** [comment]
OK, @patrikhuber is now a reviewer

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01262/joss.01262/10.21105.joss.01262.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
> @Kevin-Mattheus-Moerman I don't mind personally. But if there's anything more that would help you (like doing an official checklist), I'd be happy to do it. Should be trivial given I've already completed a pretty much full review and everything is fine from my end.
> I am on vacation though so the earliest I could do anything is the week of 12 August.

Thanks @patrikhuber I've just added you as an official reviewer. Since you did all the work and did effectively perform a review I think it is best to keep track of it officially like this. You should be able to tick the boxes at the top of this issue under your handle. 
Let me know if you have any questions. Thanks again!

**Comment by patrikhuber** [comment]
@Kevin-Mattheus-Moerman Just wanted to let you know that I still got this on my todo list, things are very busy currently. I shall get to it by end of next week, apologies it's taking a short while.

**Comment by Kevin-Mattheus-Moerman** [comment]
@patrikhuber great. Thanks for the update. I hope this can be wrapped up soon. Thanks again.

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes I noticed the "Functionality" box is not ticked. Can you summarize where you think this needs improvement or state what is needed for you to tick this box? Thanks!

**Comment by Kevin-Mattheus-Moerman** [comment]
> I see the following issues:
> 
>     * [ ]  Both `pip3 install -U Scyland3D` and `python3 setup.py install Scyland3D --user` lead to an unusable state: `ImportError: cannot import name 'main'`
> 
>     * [ ]  Calling without command line arguments does not provide help. Instead, an example directory is parsed. This violates the principle of least surprise.
> 
>     * [ ]  Argument parsing seems unnecessarily verbose. My first review suggested the `argparse` package. I'm not sure why `getopt` has been used.
> 
>     * [ ]  Image in README appears without justification. It should be used to support the initial description of what the software does.
> 
>     * [ ]  Internal comments are lacking near critical places. What, for instance, does `t = float((D - A * x - B * y - C * z) / (A * A + B * B + C * C))` do? Try to describe mathematical operations and other patches of non-contextualized code to orientate yourself and others. Particularly, [Issue 8](https://github.com/ybayle/Scyland3D/issues/8) seems disappointingly unaddressed.
> 
> 
> Other than this, the situation is improved versus the original submission; however, my sense that the software solves too specific a problem in too specific a way remains.

Are these issues resolved? Thanks.

**Comment by Kevin-Mattheus-Moerman** [comment]
@patrikhuber how are you doing? Any updates on this. Thanks again for your help!

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes can you summarize what points you feel should still be addressed? Thanks! 

> @r-barnes I noticed the "Functionality" box is not ticked. Can you summarize where you think this needs improvement or state what is needed for you to tick this box? Thanks!

**Comment by patrikhuber** [comment]
@Kevin-Mattheus-Moerman I've just done the checklist, apologies it took a while.
I haven't checked off two boxes as I am having some trouble running Scyland3D from pip - will open an issue in the repo.

Also one other thing I found, the README.md contains the sentence & link `More information can be found on the related article.` - The link is currently a 404, this needs to be updated, I guess once the paper has been accepted and published in JOSS.

**Comment by Kevin-Mattheus-Moerman** [comment]
@ybayle can you respond to the issues raised by @patrikhuber, thanks.

**Comment by Kevin-Mattheus-Moerman** [comment]
@ybayle :wave: :point_up:

**Comment by ybayle** [comment]
@Kevin-Mattheus-Moerman Hi!
Sorry for the delay. All comments are relevant and useful and are being processed. I'll push soon and keep you updated!
Thanks!

**Comment by ybayle** [comment]
@whedon set v1.0.17 as version

**Comment by whedon** [comment]
I'm sorry @ybayle, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by ybayle** [comment]
@patrikhuber Thanks for your comments. I solved ybayle/Scyland3D#10 and ybayle/Scyland3D#11. Tell me if everything is ok for you in the new version 1.0.17.

```
Also one other thing I found, the README.md contains the sentence & link More information can be found on the related article. - The link is currently a 404, this needs to be updated, I guess once the paper has been accepted and published in JOSS.
```

That's it, it will be updated upon acceptance.

@r-barnes Thanks for your suggestions that have all been added. N.B.: the use of argparse is far better than optget!

@Kevin-Mattheus-Moerman Everything is ok and up to date on my side, except for the new version. Could you update it on your side please? Thanks!

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set v1.0.17 as version

**Comment by whedon** [comment]
OK. v1.0.17 is the version.

**Comment by Kevin-Mattheus-Moerman** [comment]
@ybayle have these issues raised by @r-barnes been addressed :point_down: 

> I see the following issues:
> 
>     * [ ]  Both `pip3 install -U Scyland3D` and `python3 setup.py install Scyland3D --user` lead to an unusable state: `ImportError: cannot import name 'main'`
> 
>     * [ ]  Calling without command line arguments does not provide help. Instead, an example directory is parsed. This violates the principle of least surprise.
> 
>     * [ ]  Argument parsing seems unnecessarily verbose. My first review suggested the `argparse` package. I'm not sure why `getopt` has been used.
> 
>     * [ ]  Image in README appears without justification. It should be used to support the initial description of what the software does.
> 
>     * [ ]  Internal comments are lacking near critical places. What, for instance, does `t = float((D - A * x - B * y - C * z) / (A * A + B * B + C * C))` do? Try to describe mathematical operations and other patches of non-contextualized code to orientate yourself and others. Particularly, [Issue 8](https://github.com/ybayle/Scyland3D/issues/8) seems disappointingly unaddressed.
> 
> 
> Other than this, the situation is improved versus the original submission; however, my sense that the software solves too specific a problem in too specific a way remains.

**Comment by patrikhuber** [comment]
@ybayle @Kevin-Mattheus-Moerman I've retested today and updated the issues.

**Comment by ybayle** [comment]
> have these issues raised by @r-barnes been addressed

@Kevin-Mattheus-Moerman Yes but both reviewers raised new issues that I'm tackling.

**Comment by ybayle** [comment]
@Kevin-Mattheus-Moerman I have addressed all reviewers comments, suggestions and fixed every bug raised by the reviewers. I have added a new release on the main repo, could you update again the version to v1.0.18 please? Thanks! I'm waiting for confirmation from @r-barnes and @patrikhuber that everything is ok for the v1.0.18 release.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set v1.0.18 as version

**Comment by whedon** [comment]
OK. v1.0.18 is the version.

**Comment by patrikhuber** [comment]
Just also mentioning https://github.com/ybayle/Scyland3D/issues/13 here as it is/was part of this review.

I would now consider all my issues closed and my review done.

Note that one thing I haven't done is to verify that the output of the processing is actually correct, i.e. that the reversed landmarks are actually correctly reversed or the re-odered ones correctly re-ordered. This would require more work and writing some plotting/visualisation scripts. There are unit tests in Scyland3D for these functions though and these pass on my machine.

I also still have my original comment that the software is extremely specific, both in what it does as well as the domain that it applies to. But that's not something that can be changed, it is what it is. I believe it is the Editor's decision whether this fits into JOSS.

Let me know if anything else is needed from my side.

All the best!

**Comment by ybayle** [comment]
Thank you @patrikhuber for all your helpful comments about the code and the documentation that helped improve a lot our submission.
@Kevin-Mattheus-Moerman: I'm waiting for the validation of @r-barnes for my fixes on the last two issues he raised.

**Comment by ybayle** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01262/joss.01262/10.21105.joss.01262.pdf)

**Comment by ybayle** [comment]
@Kevin-Mattheus-Moerman I have pushed v1.0.19 on pypi that fix the issues from @patrikhuber. Could you update the version accordingly?

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set v1.0.19 as version

**Comment by whedon** [comment]
OK. v1.0.19 is the version.

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes can you check the fixed implemented? Thanks

**Comment by danielskatz** [comment]
👋 @r-barnes - note that this submission is waiting on your check

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes when you are ready can you check the fixes implemented. Let me know if you need more time. Let us know if anything else is stopping you from ticking that last box. Thanks! :christmas_tree:

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes best wishes for 2020! :tada: 
Please can you check the above. Thank!

**Comment by Kevin-Mattheus-Moerman** [comment]
@r-barnes @patrikhuber can you pick up the review or summarize why you feel the last box is  unticked? Thanks.

**Comment by patrikhuber** [comment]
@Kevin-Mattheus-Moerman I believe my comments in this thread sum it up pretty nicely, particularly my last one [[link](https://github.com/openjournals/joss-reviews/issues/1262#issuecomment-549161575)] should answer your question.

**Comment by r-barnes** [comment]
@Kevin-Mattheus-Moerman : Things look largely good to me. The software/repo seem to be in good shape. The readability and documentation of the code is sufficient. The pip installation works. I remain concerned about the specificity of the software, but leave that to the editors.

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @r-barnes @patrikhuber for your additional comments and all the help with this review! I understand the concerns about the specificity of the software. Our [submission requirements](https://joss.readthedocs.io/en/latest/submitting.html#submission-requirements) do not currently list restrictions in relation to how extremely specific/niche software might be. We've had to reject submission in the past for being too specific/niche but only because they were so exotic or specialized that we were unable to find any suitable reviewers. In this case this what not the problem. I suppose perhaps if software is too specific it might not be classified as a "significant contribution" (which is in the guidelines). However in this case I feel this software offers a significant enough contribution to warrant publication. We may reflect on our submission requirements/guidelines though to perhaps include statements on how "specific/niche" software is. 
Thanks again for your valuable contributions here!

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01262/joss.01262/10.21105.joss.01262.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@ybayle I will now proceed to process this submission for acceptance in JOSS. 

At this point can you:
- [ ] Archive the software on ZENODO and report back the DOI of the archived version here? Please ensure the title and author list of the ZENODO archived version match those of the paper (you may need to edit this information manually) .
- [ ] Can you inform me of the version tag that matches that archived version? Is it still at `v1.0.19`? 
- [ ] This work is about to be accepted in JOSS. Can you please proofread the paper yourself one last time? 

About the paper:
- [ ] the paper contains both `csv` (e.g. in `... to a csv format`) and `CSV`. Can you check you use `CSV` consistently throughout the text. 
- [ ] In use `noise` not `noises` in `...may bring noises in the results.`
- [ ] Change `Presence of duplicates landmark...` to `Presence of duplicate landmarks...`
- [ ] Rephrase `...the two reviewers that helped improved a lot the code and the documentation.` to `...the two reviewers who helped improve the code and the documentation a lot.`

**Comment by ybayle** [comment]
Thank you for all your helpful comments. Before archiving the repo we'll update the last comments on the article and the code. We'll keep you updated.

**Comment by ybayle** [comment]
We proofread the paper again and generated a new version with the latest fixes: v1.1.0 cf https://pypi.org/project/Scyland3D/1.1.0/.
The Zenodo DOI is https://doi.org/10.5281/zenodo.3637546 but is not accessible yet. It's the first time that I used Zenodo, do I have to wait few days before Zenodo displays correctly every information?

**Comment by ybayle** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01262/joss.01262/10.21105.joss.01262.pdf)

**Comment by ybayle** [comment]
@Kevin-Mattheus-Moerman 
The generated article proof and the Zenodo are ok. What is the next step?

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set 10.5281/zenodo.3637546 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3637546" target="_blank">10.5281/zenodo.3637546</a> is the archive.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set v1.1.0 as version

**Comment by whedon** [comment]
OK. v1.1.0 is the version.

**Comment by Kevin-Mattheus-Moerman** [comment]
@ybayle looks like we are all set.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.4404/hystrix-24.1-6283 is OK
- 10.11646/zootaxa.4461.1.8 is OK
- 10.1186/s12862-015-0285-5 is OK
- 10.1371/journal.pone.0206984 is OK
- 10.1111/zoj.12403 is OK
- 10.1016/B978-012778460-1/50007-7 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@openjournals/jose-eics this paper is ready to be processed

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1281

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1281, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@openjournals/joss-eics this paper is ready to be processed

**Comment by kthyng** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1226249668893585409) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1284
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01262](https://doi.org/10.21105/joss.01262)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by kthyng** [comment]
Congratulations to @ybayle on your new publications! Thanks for editor @Kevin-Mattheus-Moerman and to reviewers @r-barnes and @patrikhuber for your time and expertise.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01262/status.svg)](https://doi.org/10.21105/joss.01262)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01262">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01262/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01262/status.svg
   :target: https://doi.org/10.21105/joss.01262
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01262/status.svg)](https://doi.org/10.21105/joss.01262)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by ybayle** [comment]
@kthyng @Kevin-Mattheus-Moerman @r-barnes @patrikhuber Thank you very much for your help!

**Comment by patrikhuber** [comment]
@ybayle Congratulations on the acceptance!
